olp-cpp-sdk-authentication/src/AuthenticationClient.cpp

```diff
@@ -943,6 +1025,76 @@ client::CancellationToken AuthenticationClient::Impl::IntrospectApp(
                  std::move(introspect_app_task), std::move(wrap_callback));
 }
 
+client::CancellationToken AuthenticationClient::Impl::Authorize(
+    std::string access_token, AuthorizeRequest request,
+    AuthorizeCallback callback) {
+  using ResponseType = client::ApiResponse<AuthorizeResult, client::ApiError>;
+
+  auto task = [=](client::CancellationContext context) -> ResponseType {
+    if (!settings_.network_request_handler) {
+      return client::ApiError({static_cast<int>(http::ErrorCode::IO_ERROR),
+                               "Can not send request while offline"});
+    }
+    client::AuthenticationSettings auth_settings;
+    auth_settings.provider = [&access_token]() { return access_token; };
+
+    client::OlpClientSettings settings;
+    settings.network_request_handler = settings_.network_request_handler;
+    settings.task_scheduler = settings_.task_scheduler;
+    settings.proxy_settings = settings_.network_proxy_settings;
+    settings.authentication_settings = auth_settings;
+
+    client::OlpClient client;
+    client.SetBaseUrl(settings_.token_endpoint_url);
+    client.SetSettings(settings);
+
+    auto http_result = client.CallApi(kDecisionEndpoint, "POST", {}, {}, {},
+                                      generateAuthorizeBody(request),
+                                      kApplicationJson, context);
+
+    rapidjson::Document document;
+    rapidjson::IStreamWrapper stream(http_result.response);
+    document.ParseStream(stream);
+    if (http_result.status != http::HttpStatusCode::OK) {
+      // HttpResult response can be error message or valid json with it.
+      std::string msg = http_result.response.str();
+      if (!document.HasParseError() && document.HasMember(Constants::MESSAGE)) {
+        msg = document[Constants::MESSAGE].GetString();
+      }
+      return client::ApiError({http_result.status, msg});
+    } else if (!document.HasParseError() &&
+               document.HasMember(Constants::ERROR_CODE) &&
+               document[Constants::ERROR_CODE].IsInt()) {
+      std::string msg =
+          "Error code:" +
+          std::to_string(document[Constants::ERROR_CODE].GetInt());
+      return client::ApiError(
+          {static_cast<int>(http::ErrorCode::UNKNOWN_ERROR), msg});
```

Comment:
But this is not very good that we do not return the complete error to the user.

Reply:
added error message, if exist.

Reply:
I meant something else, but for now lets keep like this.
The idea is that the AuthorizeResult inherits from BaseResult which includes all necessary fields and parsing logic.